[Watcher] migrate to rtl

kapral18 opened the pull request:
Closes #241536
Fixes #163531
Fixes #163532

## Summary
- Replace Watcher’s Enzyme-based Jest coverage with React Testing Library (RTL).
- Remove snapshots and legacy testbed/page helper modules.
- Add plugin-local JSDOM mocks for `@kbn/code-editor` and `@elastic/charts`.

## Notes for reviewers
- In `x-pack/platform/plugins/private/watcher/__jest__/client_integration/watch_create_json_page.test.tsx`, the “results flyout” assertions were intentionally consolidated:
  - Main tests:
    - `should set the correct condition met status`
    - `should set the correct values in the table`
    → migrated into: `test.each([...])('should render correct table status for $actionMode (conditionMet=$conditionMet)', ...)`
  - Main tests:
    - `flyout renders`
    - `condition status is not displayed`
    → migrated into: `test('when API returns no results, flyout renders and condition status is not displayed', ...)`
    
### Table of Changes

https://gist.github.com/kapral18/115d9f0507c4d46e118ccb37f269c847

## Test plan
- [x] node scripts/eslint x-pack/platform/plugins/private/watcher
- [x] node scripts/type_check --project x-pack/platform/plugins/private/watcher/tsconfig.json
- [x] node scripts/jest x-pack/platform/plugins/private/watcher/__jest__

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/377428)
* Commit: def241387d28976544acea204d12d3e372e60314

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/377428#019b7b7f-7275-42c1-864d-a8fdb30dc7e1)

### History
* :broken_heart: [Build #377413](https://buildkite.com/elastic/kibana-pull-request/builds/377413) failed 1b6009d4609a2002a0feed11f394d699b02d70d8

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377428","number":377428,"commit":"def241387d28976544acea204d12d3e372e60314"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377413","number":377413,"commit":"1b6009d4609a2002a0feed11f394d699b02d70d8"}],"number":377428}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/377429)
* Commit: 815e10ee82203acb90d2361dc62b8bee0bd8cf22

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Count of Enzyme imports
> Enzyme is no longer supported, and we should switch to @testing-library/react instead.


| id | [before](https://github.com/elastic/kibana/commit/5d3ae06d5cab3b983da35e6275b3c595832a686a) | [after](https://github.com/elastic/kibana/commit/815e10ee82203acb90d2361dc62b8bee0bd8cf22) | diff |
| --- | --- | --- | --- |
| `watcher` | 1 | 0 | -1 |

<details><summary>Unknown metric groups</summary>

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/5d3ae06d5cab3b983da35e6275b3c595832a686a) | [after](https://github.com/elastic/kibana/commit/815e10ee82203acb90d2361dc62b8bee0bd8cf22) | diff |
| --- | --- | --- | --- |
| `watcher` | 26 | 2 | -24 |


</details>


### History
* :broken_heart: [Build #377428](https://buildkite.com/elastic/kibana-pull-request/builds/377428) failed def241387d28976544acea204d12d3e372e60314
* :broken_heart: [Build #377413](https://buildkite.com/elastic/kibana-pull-request/builds/377413) failed 1b6009d4609a2002a0feed11f394d699b02d70d8

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377429","number":377429,"commit":"815e10ee82203acb90d2361dc62b8bee0bd8cf22"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377428","number":377428,"commit":"def241387d28976544acea204d12d3e372e60314"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/377413","number":377413,"commit":"1b6009d4609a2002a0feed11f394d699b02d70d8"}],"number":377429}
buildkite-pr-comment-->

kapral18 committed:
chore(watcher): migrate jest tests to RTL

- Migrate Watcher client integration Jest tests from Enzyme/testbed to RTL
- Remove legacy testbed helper modules and snapshot usage
- Add plugin-local mocks for @kbn/code-editor and @elastic/charts
- Update Watcher test infra (tsconfig, moon.yml, http mocks)

kapral18 commented:
> > Not sure I get this do you mean screenshots of before and after test runs or a markdown file or something?
> 
> Yes, screenshots or whatever makes the most sense.

Yeah I tried both markdown and screenshots, nothing made sense. Github squeezes the markdown makes it impossible to render. 

The only good solution I found was generate a csv with pointers. That's what I did. Check updated PR description for Table of changes.

mattkime commented on x-pack/platform/plugins/private/watcher/__mocks__/@elastic/charts/index.tsx:
```diff
@@ -0,0 +1,27 @@
+ * - Avoid RAF (requestAnimationFrame) open-handle leaks in Jest.
+ * - Provide stub exports for the chart components used by Watcher.
+ */
+export const AnnotationDomainType = { YDomain: 'YDomain' };
```
Still would be good to add comments stating how they're being used

kapral18 replied:
How they are being used is describe in https://jestjs.io/docs/manual-mocks. This is not the first usage of manual mocks in kibana tests. Similarly to how we don't add links to less popular react api usage, I don't think we should add this either. The assumption is when you read jest or other similar core libraries like react etc, you either know their mechanics or are capable of looking them up. Adding comments to basic jest mechanics will imho only contribute to comment pollution. Wdyt?

kapral18 committed:
chore: remove unndeeded charts mock

kapral18 committed:
test: reuse existing code-editor mocks

mattkime reviewed (approved):
lookin' good!
